accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/authorizableinstaller/impl/PrefetchedAuthorizablesUserManager.java

```diff
@@ -0,0 +1,125 @@
+package biz.netcentric.cq.tools.actool.authorizableinstaller.impl;
+
+import static biz.netcentric.cq.tools.actool.history.PersistableInstallationLogger.msHumanReadable;
+
+import java.security.Principal;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import javax.jcr.RepositoryException;
+
+import org.apache.jackrabbit.api.security.user.Authorizable;
+import org.apache.jackrabbit.api.security.user.Group;
+import org.apache.jackrabbit.api.security.user.Query;
+import org.apache.jackrabbit.api.security.user.QueryBuilder;
+import org.apache.jackrabbit.api.security.user.User;
+import org.apache.jackrabbit.api.security.user.UserManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import biz.netcentric.cq.tools.actool.history.InstallationLogger;
+
+/** Prefetches all authorizables in constructor to be able to answer getAuthorizable(userId) requests much quicker than the ootb oak
+ * UserManager itself. */
+class PrefetchedAuthorizablesUserManager implements UserManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PrefetchedAuthorizablesUserManager.class);
+
+    static final Query ALL_AUTHORIZABLES_QUERY = new Query() {
+        public <T> void build(QueryBuilder<T> builder) {
+            // fetch all authorizables
+        }
+    };
+
+    private final UserManager delegate;
+    private final Map<String, Authorizable> authorizableCache;
+
+    public PrefetchedAuthorizablesUserManager(UserManager delegate, InstallationLogger installLog) throws RepositoryException {
+        this.delegate = delegate;
+        this.authorizableCache = new HashMap<>();
+
+        long startPrefetch = System.currentTimeMillis();
+        Iterator<Authorizable> allAuthorizables = delegate.findAuthorizables(ALL_AUTHORIZABLES_QUERY);
```

Comment:
Prefetching all authorizables (using one query under the hood) turns out to be a lot more efficient than running many individual getAuthorizable() calls

Reply:
Do we know why? I once figured out in https://projects.netcentric.biz/wiki/display/CQ/Authorizable+vs.+Principal
> UserManager.getAuthorizable(String id)
This method is usually very fast because no query is involved because the node's jcr:uuid property (i.e. the nodes ID) is equal to UUID.nameUUIDFromBytes(name.getBytes(StandardCharsets.UTF_8)).
Therefore, Session.getNodeByIdentifier(...)is used to quickly retrieve the underlying node.
That is the case for all authorizables being created via the UserManager!

Is that observation no longer correct?

Reply:
So I just did measurements with a ~4000 authorizables setup and caching it like this brings down the installation time significantly. However I found that my current project's performance was improved but  @tomasalmeida 's problem  was not yet solved (~100.000 users in getDeclaredMembers()) - hence I followed this path further, see https://github.com/Netcentric/accesscontroltool/blob/1413e9315c2d3199e8f52dda087def95c39c3802/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/authorizableinstaller/impl/PrefetchingUserManager.java#L40

Reply:
I deployed this code and tested it against an old version. Comparison values:

`time <deploy command>`

actool 2.3.2 
real	2m43.451s
real	2m45.496s
real	3m5.596s
real	2m57.520s

this branch (installed over the 2.3.2)
real	0m22.456s
real	0m18.718s
real	0m18.224s

I have in my local server more than 150000 users.